pkg/scheduler/queue/queue.go

```diff
@@ -151,6 +152,9 @@ func (q *RequestQueue) dispatcherLoop() {
 			case notifyShutdown:
 				queues.notifyQuerierShutdown(qe.querierID)
 				needToDispatchQueries = true
+
+				// Tell any waiting GetNextRequestForQuerier calls for this querier that nothing is coming.
+				q.cancelWaitingConnectionsForQuerier(qe.querierID, waitingQuerierConnections)
```

Comment:
The shutdown notification is a best effort. It's not guaranteed to be received (e.g. if the querier OOMs or the K8S node where it was running gets abruptly terminated).

Reply:
If the shutdown notification is not received, we'll try to send a request to the querier (in `Scheduler.QuerierLoop()`), get an error, call `unregisterConnection ` and not call `GetNextRequestForQuerier` again - so this is OK.

Reply:
What you're describing is the logic in this PR, not `main`, right? Cause I can see what you describe happens with the change proposed in this PR.

Reply:
I find it a bit difficult to follow this logic. Wouldn't be safer and more clear if we remove a querier from `waitingQuerierConnections` even when handling `unregisterConnection`?

Reply:
> Just to be on the same page: what you're describing is the logic in this PR, not `main`, right? Cause I can see what you describe happens with the change proposed in this PR.

The behaviour in the unclean shutdown case should be the same in both.

> I find it a bit difficult to follow this logic. Wouldn't be safer and more clear if we remove a querier from `waitingQuerierConnections` even when handling `unregisterConnection`?

`unregisterConnection` is only called by `Scheduler.QuerierLoop()`, and so isn't called concurrently with `GetNextRequestForQuerier`, so removing the pending connection from `waitingQuerierConnections` is not necessary, because it won't be there.

Reply:
I would improve the comment to say that this is just a nice to have, but it's OK if we don't receive the shutdown notifications and `waitingQuerierConnections` will be cleaned up when trying to dispatch a request to the querier the next time.